### PR TITLE
udev/net: introduce WakeOnLanPassword=

### LINKED ISSUE #20913: How to specify the SecureOn password in a Link file?
A link file allows configuring the WakeOnLan policy of network interface with the `WakeOnLan=` setting.
The manual page [systemd.link(5)](https://www.freedesktop.org/software/systemd/man/systemd.link.html) says it's also possible to enable a password with `WakeOnLan=secureon`, however it doesn't mention how to specify the password itself. How should I configure the password?

### PR DESCRIPTION
**Pull request by yuwata**
Closes #20913.

### CODE REVIEW

File: man/systemd.link.xml [hunk 1]
```diff
@@ -548,6 +548,15 @@
           cleared.</para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>WakeOnLanSecureOnPassword=</varname></term>
+        <listitem>
+          <para>Specifies the SecureOn(tm) password for MagicPacket(tm). Takes 6 bytes in Ethernet MAC
+          hex format (<literal>aa:bb:cc:dd:ee:ff</literal>). This will be ignored when
+          <varname>WakeOnLan=</varname> is specified without <literal>secureon</literal>. Defaults to
+          unset, and the current value will not be changed.</para>
```
Comment (poettering) [inline man/systemd.link.xml]: it's so weird that they use MAC addresses as passwords.

File: man/systemd.link.xml [hunk 2]
```diff
@@ -548,6 +548,15 @@
           cleared.</para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>WakeOnLanSecureOnPassword=</varname></term>
```
Comment (poettering) [inline man/systemd.link.xml]: i'd probably just name this "WakeOnLanPassword=". i.e. no need to repeat the trademark here, the generic naming as just "password" is clear enough.

File: man/systemd.link.xml [hunk 3]
```diff
@@ -548,6 +548,15 @@
...
+        <listitem>
+          <para>Specifies the SecureOn(tm) password for MagicPacket(tm). Takes 6 bytes in Ethernet MAC
+          hex format (<literal>aa:bb:cc:dd:ee:ff</literal>). This will be ignored when
+          <varname>WakeOnLan=</varname> is specified without <literal>secureon</literal>. Defaults to
+          unset, and the current value will not be changed.</para>
+        </listitem>
+      </varlistentry>
```
Comment (poettering) [inline man/systemd.link.xml:572]: I think we should also make sure that if someone sets WakeOnLanPassword= but doesn't touch WakeOnLan= at all that we imply WakeOnLan=secureon.

**Review by poettering — CHANGES_REQUESTED** [review]
looks great. 

(I think eventually we should make this work a bit more like the wireguard key handling, i.e. allow specifiying a path in the fs to read the password from, and support AF_UNIX sockets for that too, i.e. wireguard_read_key_file())

File: src/shared/ethtool-util.c [hunk 4]
```diff
@@ -1359,6 +1374,48 @@ int config_parse_wol(
...
+                void *data,
+                void *userdata) {
+
+        _cleanup_(erase_and_freep) struct ether_addr *p = NULL;
+        struct ether_addr **password = data;
+        int r;
+
+        assert(filename);
+        assert(rvalue);
+
```
Comment (bluca) [inline src/shared/ethtool-util.c]: `assert(password)`

**Comment by rnhmjoj** [comment]
> (I think eventually we should make this work a bit more like the wireguard key handling, i.e. allow specifiying a path in the fs to read the password from

I was going to ask the same thing: it's better not to mix secrets and configuration, if possible. This is particularly probematic in NixOS where the (generated) configuration files are world-readable by default.

Anyway, I'll try test this patch as soon as I have time and report back. 
Thank you for working on this, @yuwata.

**Comment by poettering** [comment]
(I think eventually we should also hook up the creds logic to wireguard and secureon, i.e. so that you can do LoadCredentials= in the networkd unit file to set these, or even inherit things down from nspawn params, and so on. i.e. try read_credential() on some name derived from the network config file name or so, and mayba single fallback to a fixed string)

**Comment by yuwata** [comment]
@poettering, @bluca, and @rnhmjoj Thank you for your review and comments. Updated.
- Renamed WakeOnLanSecureOnPassword= -> WakeOnLanPassword=
- The setting now also support a boolean string and a path to a file which contains the password.

PTAL.

(I will address wireguard thing in other later PR.)

File: man/systemd.link.xml [hunk 5]
```diff
@@ -537,8 +537,8 @@
             <varlistentry>
               <term><option>secureon</option></term>
               <listitem>
-                <para>Enable secureon(tm) password for MagicPacket(tm).
-                </para>
+                <para>Enable SecureOn(tm) password for MagicPacket(tm). Implied when
```
Comment (poettering) [inline man/systemd.link.xml]: btw, i wonder about the (tm) thing. I can't find the specs for these two concepts, and whether they carry the TM thing there. Are these really trademarks? or just cargoculted from ethtool?

A the very least this should be a unicode ™, of course ;-)

File: src/shared/ethtool-util.c [hunk 6]
```diff
@@ -400,7 +400,12 @@ int ethtool_get_permanent_macaddr(int *ethtool_fd, const char *ifname, struct et
                 dest = _v;                             \
         } while(false)
 
-int ethtool_set_wol(int *ethtool_fd, const char *ifname, uint32_t wolopts) {
+int ethtool_set_wol(
+                int *ethtool_fd,
+                const char *ifname,
+                uint32_t wolopts,
+                const uint8_t *password) {
```
Comment (poettering) [inline src/shared/ethtool-util.c]: maybe change last arg to: `uint8_t password[SOPASS_MAX]` to make clear what is expected here?

File: src/udev/net/link-config.c [hunk 7]
```diff
@@ -126,6 +127,32 @@ static int link_parse_wol_password(LinkConfig *link, const char *str) {
...
+        if (r < 0)
+                return r;
+
+        cred_name = strjoin(base, ".wol.password");
+        if (!cred_name)
+                return r;
+
+        r = read_credential(cred_name, (void**) &password, NULL);
+        if (r < 0)
+                return r;
```
Comment (poettering) [inline src/udev/net/link-config.c:179]: one idea: if this returns ENOENT, maybe fallback to `read_credential("wol.password", …)`, i.e. without the basename prefix. I figure some people might not be interested to set a per-link wol password,  but only a fixed one that applies to all links

File: man/systemd.link.xml (hunk 1 above)
Comment (yuwata) [reply man/systemd.link.xml]: Slightly rephrased the explanation.

**Comment by yuwata** [comment]
Oh, udevd is coredumped on focal-s390x. That may be related. I will rework on that.

**Comment by yuwata** [comment]
I hope now the segfault is fixed.

File: man/systemd.link.xml (hunk 5 above)
Comment (poettering) [reply man/systemd.link.xml]: I am having trouble finding any spec for these techs. All I found is this:

https://www.amd.com/system/files/TechDocs/20213.pdf

And that doc never puts a TM or so on the MagicPacket thing. SecureOn is not mentioned at all there.

File: src/udev/net/link-config.c [hunk 8]
```diff
@@ -834,6 +867,14 @@ int config_parse_wol_password(
         assert(userdata);
 
         if (isempty(rvalue)) {
+                link->wol_password_from_cred = false;
+                link->wol_password = erase_and_free(link->wol_password);
+                return 0;
+        }
+
+        r = parse_boolean(rvalue);
```
Comment (poettering) [inline src/udev/net/link-config.c]: hmm, this is definitely a workable approach, but i wonder if it wouldn't be simpler and more natural to simply say that if `WakeOnLan=secureon` is selected but `WakeOnLanPassword=` is not we'll search for the password in the creds? And then not have any boolean parsing here at all?

That way: 

`WakeOnLan=secureon` without `WakeOnLanPassword=` → search for the password in the creds and use it if it exists
`WakeOnLanPassword=` without `WakeOnLan=` → imply `WakeOnLan=secureon`

File: src/shared/ethtool-util.c [hunk 9]
```diff
@@ -425,16 +431,37 @@ int ethtool_set_wol(int *ethtool_fd, const char *ifname, uint32_t wolopts) {
...
+        }
 
+        r = 0;
         ecmd.cmd = ETHTOOL_SWOL;
         if (ioctl(*ethtool_fd, SIOCETHTOOL, &ifr) < 0)
-                return -errno;
+                r = -errno;
 
-        return 0;
+        zero(ecmd);
```
Comment (poettering) [inline src/shared/ethtool-util.c]: Should use explicit_bzero_safe(), zero() is not enough compiler might optimize that away

**Review by poettering — CHANGES_REQUESTED** [review]

File: man/systemd.link.xml (hunk 5 above)
Comment (poettering) [reply man/systemd.link.xml]: (i'd probably just drop the TM though)

File: man/systemd.link.xml [hunk 10]
```diff
@@ -548,6 +557,17 @@
...
       </varlistentry>
+      <varlistentry>
+        <term><varname>WakeOnLanPassword=</varname></term>
+        <listitem>
+          <para>Specifies the SecureOn password for MagicPacket. Takes an absolute path to a file which
+          contains the password, or the plain password. The password in the file, or specified in this
+          option must be 6 bytes in hex format with each byte separated by a colon
+          (<literal>:</literal>) like an Ethernet MAC address, e.g.,
+          <literal>aa:bb:cc:dd:ee:ff</literal>. This implies <varname>WakeOnLan=secureon</varname>.
+          Defaults to unset, and the current value will not be changed.</para>
```
Comment (poettering) [inline man/systemd.link.xml:570]: maybe also mention that the the absolute path may optionally refer to an AF_UNIX/SOCK_STREAM socket, in which case a connection is made to it and the password read off the connection.

**Review by poettering — CHANGES_REQUESTED** [review]
looks excellent!

**Pull request merged**